### ISSUE TITLE
Mixed precision parity: Float8 types, LayerPrecisionPolicy, generic LossScaler<T>, and richer AutocastScope

### ISSUE THREAD

**Issue by ooples** [ooples/AiDotNet.Tensors#197]
## Context

`AiDotNet.Tensors` 0.46.0 ships `AutocastScope` + `GradScaler` for mixed-precision training. The consuming library (`AiDotNet`) currently carries its own `src/MixedPrecision/` folder (~2,500 LOC) because these Tensors primitives, while good, are too narrow for PyTorch-parity use. We want to consolidate on Tensors as the single source of truth — but that means Tensors needs to cover the full surface. This issue is the spec for the missing pieces.

### 1. FP8 types — `Float8E4M3`, `Float8E5M2`

AiDotNet has a 627-line `Float8Types.cs` providing:
- `readonly struct Float8E4M3 : IEquatable<Float8E4M3>, IComparable<Float8E4M3>` — 1 sign, 4 exp, 3 mantissa (NVIDIA H100 forward pass default)
- `readonly struct Float8E5M2 : IEquatable<Float8E5M2>, IComparable<Float8E5M2>` — 1 sign, 5 exp, 2 mantissa (NVIDIA H100 backward pass default)
- `Float8Extensions` — conversions to/from `float`, `Half`, `double`; element-wise kernels
- NaN / infinity handling matching IEEE-754 finite-only FP8 semantics
- Range clamping for overflow saturation (OCP FP8 spec)

Place under `AiDotNet.Tensors.NumericOperations`. Register with `INumericOperations<T>` so generic pipelines can use `Float8E4M3`/`E5M2` as T.

**Acceptance tests:** roundtrip fp32 → fp8 → fp32 within expected error for representable range; NaN+Inf propagation matches OCP spec; saturating overflow to max-representable on inputs beyond range.

### 2. `LayerPrecisionPolicy`

Per-layer precision control. Currently in `AiDotNet.MixedPrecision.LayerPrecisionPolicy`. Spec:

\`\`\`csharp
public class LayerPrecisionPolicy
{
    public PrecisionMode DefaultPrecision { get; }
    public LayerPrecisionPolicy(PrecisionMode defaultPrecision = PrecisionMode.Float16);
    public LayerPrecisionPolicy SetPrecision(string exactLayerName, PrecisionMode precision);
    public LayerPrecisionPolicy AddPattern(string substringPattern, PrecisionMode precision); // case-insensitive substring match
    public LayerPrecisionPolicy KeepInFP32(string pattern);          // syntactic sugar
    public PrecisionMode GetLayerPrecision(string layerName);
    public bool ShouldSkipMixedPrecision(string layerName);          // returns true if layer's policy is FP32/None
    public static LayerPrecisionPolicy ForFP16();                    // default: normalization, softmax, loss → FP32; rest → FP16
    public static LayerPrecisionPolicy ForBF16();
    public static LayerPrecisionPolicy ForFP8();
}
\`\`\`

Place under `AiDotNet.Tensors.Engines.Gpu` next to `AutocastScope`.

**Acceptance tests:** substring pattern matches multiple layer names case-insensitively; exact-match wins over pattern; `ForFP16` default keeps BatchNorm/LayerNorm/Softmax/Loss in FP32.

### 3. `AutocastScope` — named tensor cache + layer awareness

Current `AutocastScope`:
- Static `Current`, `IsEnabled`, `ActivePrecision`
- `MaybeConvertInput(backend, buf, size)` / `MaybeConvertOutput(...)`
- `ShouldAutocast(operationName)`

Missing (ported from AiDotNet `MixedPrecisionScope`):
- Constructor overload: \`new AutocastScope(PrecisionMode precision, LayerPrecisionPolicy? policy = null)\`
- Per-name tensor cache so layers can do FP32 master-weight storage alongside FP16 compute copies:
  - `Tensor<Half> RegisterAndCastToFP16(string name, Tensor<float> fp32)`
  - `Tensor<float>? GetFP32Tensor(string name)`
  - `Tensor<Half>? GetFP16Tensor(string name)`
  - `bool HasTensor(string name)`
  - `void ClearTensors()`
- `bool ShouldUseFP32(string layerName)` — consults the policy
- `bool ShouldUseHigherPrecision(string layerName)`
- Static helpers: `CastToFP32(Tensor<Half>)` / `CastToFP16(Tensor<float>)` — pure-CPU paths for testing
- Nested-scope support: instance field \`_previous\` restored on Dispose so nested \`using\` scopes work

### 4. `GradScaler` — generic over `T`

Current `GradScaler`:
- Non-generic, float-only based on signatures
- `ScaleLoss(Tensor<T> loss, IEngine)` / `Unscale(Tensor<T>[], IEngine)` / `Update()` / `ShouldStep()` / `FoundInfOrNan` / `Scale`

Missing (ported from AiDotNet `LossScaler<T>`):
- Make generic: \`public sealed class GradScaler<T>\`
- Tunable properties: \`bool DynamicScaling\`, \`int GrowthInterval\`, \`double GrowthFactor\`, \`double BackoffFactor\`, \`double MinScale\`, \`double MaxScale\`
- \`Vector<T>\` overload: \`void Unscale(Vector<T>[] gradients, IEngine)\`
- Scalar overload: \`T ScaleLoss(T loss)\` / \`T UnscaleGradient(T gradient)\`
- Overflow detection: \`bool HasOverflow(T value)\`, \`bool DetectOverflow(Tensor<T>)\`, \`bool DetectOverflow(Vector<T>)\`
- Atomic operation: \`bool UnscaleGradientsAndCheck(Tensor<T> grads)\` — returns true if step should be skipped due to inf/nan
- \`void Reset(double? newInitialScale = null)\`

**Acceptance tests:** growth after N successful steps; backoff on overflow; min/max scale clamping; generic dispatch for `float`, `double`, `Half`, `Float8E4M3` (from sub-feature 1).

### 5. `MixedPrecisionContext` — training orchestration

Wraps AutocastScope + GradScaler into one object so training loops can be written as:

\`\`\`csharp
using var ctx = new MixedPrecisionContext<T>(
    policy: LayerPrecisionPolicy.ForFP16(),
    initialLossScale: 65536.0);

foreach (var (x, y) in batches)
{
    using (ctx.BeginAutocast())
    {
        var logits = model.Forward(x);
        var loss = lossFn(logits, y);
        var scaledLoss = ctx.Scaler.ScaleLoss(loss);
        scaledLoss.Backward();
    }

    if (ctx.Scaler.UnscaleGradientsAndCheck(gradients))
        optimizer.Step(gradients, learningRate);
    ctx.Scaler.Update();
}
\`\`\`

### 6. `MixedPrecisionTrainingLoop<T>` — high-level helper

Bundles the above into:

\`\`\`csharp
public class MixedPrecisionTrainingLoop<T>
{
    public MixedPrecisionTrainingLoop(
        MixedPrecisionContext<T> context,
        Func<Tensor<T>, Tensor<T>> forward,
        ILossFunction<T> loss);

    public TrainingStepResult<T> Step(Tensor<T> input, Tensor<T> target, float learningRate);
}
public record TrainingStepResult<T>(T Loss, bool StepTaken, double CurrentLossScale);
\`\`\`

## Reference implementation

Full source currently at \`ooples/AiDotNet\` \`src/MixedPrecision/\`:
- \`Float8Types.cs\` (627 LOC)
- \`LayerPrecisionPolicy.cs\` (335 LOC)
- \`LossScaler.cs\` (401 LOC, will become \`GradScaler<T>\`)
- \`MixedPrecisionContext.cs\` (314 LOC)
- \`MixedPrecisionScope.cs\` (340 LOC — merges into \`AutocastScope\`)
- \`MixedPrecisionTrainingLoop.cs\` (145 LOC)
- \`MixedPrecisionConfig.cs\` (317 LOC — merges into \`TensorCodecOptions\`)

These are Apache-2.0 licensed and can be ported wholesale.

## Why this matters

PyTorch's \`torch.cuda.amp\` gives users:
- \`autocast()\` with per-layer type rules
- \`GradScaler\` with dynamic scaling
- \`bfloat16\`/\`float16\`/\`float8_e4m3fn\`/\`float8_e5m2\` types

Without these in Tensors, consumers (AiDotNet, CrowdTrainer, future callers) must either build their own (duplicated effort) or accept worse mixed-precision UX than PyTorch. This closes the gap.

## Non-goals

- Does not change \`TensorCodecOptions.EnableMixedPrecision\` semantics — that stays as a master on/off flag.
- Does not replace the existing raw GPU-buffer cast APIs \`MaybeConvertInput\` / \`MaybeConvertOutput\` — those remain for low-level callers.

## Estimated scope

~2,500 LOC port + ~800 LOC new tests. Can ship as \`0.47.0\`.
